modin/core/execution/client/container.py

```diff
@@ -0,0 +1,742 @@
+        Query compiler class to contain. Should be a subclass of BaseQueryCompiler.
+    io_class : type
+        The IO class to use for reading and writing data. Should be a subclass
+        of modin.core.io.io.BaseIO.
+    """
+
+    def __init__(self, query_compiler_class: type, io_class: type):
+        self._qc = {}
+        self._query_compiler_class = query_compiler_class
+        self._io_class = io_class
+
+    def _generate_id(self) -> UUID:
+        """
+        Generate an ID for a new query compiler.
+
+        Returns
+        -------
+        UUID
+            The generated ID.
+        """
+        id = uuid4()
+        while id in self._qc:
+            id = uuid4()
+        return id
+
+    def add_query_compiler(self, qc: BaseQueryCompiler) -> UUID:
+        """
+        Add a query compiler to the container.
+
+        Parameters
+        ----------
+        qc : BaseQueryCompiler
+
+        Returns
+        -------
+        UUID
+            The ID of the query compiler.
+        """
+        id = self._generate_id()
+        self._qc[id] = qc
+        return id
+
+    def to_pandas(self, id: UUID) -> pandas.DataFrame:
+        """
+        Convert the query compiler to a pandas DataFrame.
+
+        Parameters
+        ----------
+        id : UUID
+            The ID of the query compiler to convert.
+
+        Returns
+        -------
+        pandas.DataFrame
+            The converted DataFrame.
+        """
+        return self._qc[id].to_pandas()
+
+    class DefaultToPandasResult(NamedTuple):
+        """
+        The result of ``default_to_pandas``.
+
+        Parameters
+        ----------
+        result : Any
+            The result of the operation.
+        result_is_qc_id : bool
+            Whether the result is a query compiler ID.
+        """
+
+        result: Any
+        result_is_qc_id: bool
+
+    def default_to_pandas(
+        self, id: UUID, pandas_op: Union[str, callable], *args: Any, **kwargs: dict
+    ) -> DefaultToPandasResult:  # noqa: D401
+        """
+        Default to pandas for an operation on a query compiler.
+
+        Use the inner query compiler's default_to_pandas to execute the
+        operation on a pandas dataframe.
+
+        Parameters
+        ----------
+        id : UUID
+            The ID of the query compiler.
+        pandas_op : Union[str, callable]
+            The operation to perform.
+        *args : iterable
+            The arguments to pass to the operation.
+        **kwargs : dict
+            The keyword arguments to pass to the operation.
+
+        Returns
+        -------
+        DefaultToPandasResult
+            The result of the operation. The result is a query compiler ID if
+            and only if the result of the pandas operation is a new
+            query compiler.
+        """
+        result = self._qc[id].default_to_pandas(pandas_op, *args, **kwargs)
+        result_is_qc_id = isinstance(result, self._query_compiler_class)
```

Comment:
Is `result_is_qc_id` the best name for when we are checking if the types match? The documentation says that the result is a qc ID iff the result is a new query compiler. Doesn't this mean we need to check and see if the qcs are the same? Or do they just have to be the same class?